fbgemm_gpu/hip_kernel/split_tbe_fwd_hip.cpp

```diff
@@ -264,18 +323,45 @@ __device__ void split_tbe_forward_unweighted_hip_kernel(
     int32_t length_mod = length & length_mask;
 
     int itr = 0;
-    if(length_mod == 0)
-        goto L_end;
+    if(length_mod != 0){
 
-    #pragma unroll
-    for(int i=0; i < bag_unroll; i++){
-        indices[i] = p_indices[i];
-    }
+#pragma unroll
+        for(int i=0; i < bag_unroll; i++){
+            indices[i] = p_indices[i];
+        }
 
-    itr += bag_unroll;
-    p_indices += bag_unroll;
+        itr += bag_unroll;
+        p_indices += bag_unroll;
 
-    // LOOP
+        // LOOP
+        switch (emb_dim){
+            case 64:
+                LOOP_COMPUTATION(64);
+                break;
+            case 128:
+                LOOP_COMPUTATION(128);
+                break;
+            case 192:
+                LOOP_COMPUTATION(192);
+                break;
+            case 256:
+                LOOP_COMPUTATION(256);
+        }
+    }
+    switch (emb_dim){
+        case 64:
+            LAST_AND_STORE(64);
+            return;
+        case 128:
+            LAST_AND_STORE(128);
+            return;
+        case 192:
+            LAST_AND_STORE(192);
+            return;
+        case 256:
+            LAST_AND_STORE(256);
+    }
```

Comment:
the code size is enlarged 4x due to you expand everything in this loop, potentially will damage the I-CACHE and cost more power. Beside, emb=64 will lose the occupancy, because it will have the same register usage of emb=256. How would you gonna to support any other emb_dim not 64x?

Reply:
@carlushuang , I would agree that your first approach is more straightforward.

Reply:
@weihanmines to the 2 macro functions, we may inline __device__ function them, with or without inline depends on the performance.